Sources/SyncDataProviders/Settings/SettingsProvider.swift

```diff
@@ -0,0 +1,341 @@
+
+                do {
+                    let responseHandler = try SettingsResponseHandler(
+                        received: received,
+                        clientTimestamp: clientTimestamp,
+                        settingsHandlers: settingsHandlers,
+                        context: context,
+                        crypter: crypter,
+                        deduplicateEntities: isInitial
+                    )
+                    let idsOfItemsToClearModifiedAt = try cleanUpSentItems(
+                        sent,
+                        receivedKeys: Set(responseHandler.receivedByKey.keys),
+                        clientTimestamp: clientTimestamp,
+                        in: context
+                    )
+                    try responseHandler.processReceivedSettings()
+
+#if DEBUG
+                    try willSaveContextAfterApplyingSyncResponse()
+#endif
+                    let keys = idsOfItemsToClearModifiedAt.union(
+                        Set(responseHandler.receivedByKey.keys)
+                            .subtracting(responseHandler.idsOfItemsThatRetainModifiedAt)
+                    )
+                    try clearModifiedAtAndSaveContext(keys: keys, clientTimestamp: clientTimestamp, in: context)
+                    break
+                } catch {
+                    if (error as NSError).code == NSManagedObjectMergeError {
+                        context.reset()
+                        saveAttemptsLeft -= 1
+                        if saveAttemptsLeft == 0 {
+                            saveError = error
+                            break
+                        }
+                    } else {
+                        saveError = error
+                        break
+                    }
+                }
+
+            }
+        }
+        if let saveError {
+            throw saveError
+        }
+
+        if let serverTimestamp {
+            lastSyncTimestamp = serverTimestamp
+            syncDidUpdateData()
+        }
+    }
+
+    func cleanUpSentItems(
+        _ sent: [Syncable],
+        receivedKeys: Set<String>,
+        clientTimestamp: Date,
+        in context: NSManagedObjectContext
+    ) throws -> Set<String> {
+
+        if sent.isEmpty {
+            return []
+        }
+        let keys = sent.compactMap { SyncableSettingAdapter(syncable: $0).uuid }
+        let settingsMetadata = try SyncableSettingsMetadataUtils.fetchSettingsMetadata(for: keys, in: context)
+
+        var idsOfItemsToClearModifiedAt = Set<String>()
+
+        for metadata in settingsMetadata {
+            guard let handler = settingsHandlers[Setting(key: metadata.key)] else {
+                continue
+            }
+
+            if let lastModified = metadata.lastModified, lastModified > clientTimestamp {
+                continue
+            }
+            let isLocalChangeRejectedBySync: Bool = receivedKeys.contains(metadata.key)
+            let isPendingDeletion = try handler.getValue() == nil
```

Comment:
I'm not sure about this. It seems to be prone to timing issues, if user logs out while response is being processed. I know, rather rare scenario, but I wonder if we could make the state of the "pending deletion" more clear and definitive?

Reply:
Great catch. I think we can only limit the potential damage here. How about fetching values for all settings at once, right after the call to `SyncableSettingsMetadataUtils.fetchSettingsMetadata(for: keys, in: context)`? I'm not sure it would help much as the rest of this function body is really fast anyway, but it will at least look better and make the intention clear (i.e. "here is the place where we fetch all data, and later we only check values").

Reply:
I've added this change in https://github.com/duckduckgo/BrowserServicesKit/pull/485/commits/b6fa81101e37b6de4cbbc304d962a00dca247c6e. Let me know in case you have a better idea. Now that I think of it, settings could be read before creating SettingsResponseHandler (inside handleSyncResponse) to even better align the timing with the moment of creating/resetting context.

Reply:
In fact this: `if let lastModified = metadata.lastModified, lastModified > clientTimestamp` should actually be a good guard agaisnt what I wrote - that feels like the best effort to avoid issues here, given that we can't guarantee transcationality of the changes.

Reply:
If I understood your original concern correctly, this line wouldn't guard us:
* timestamps are fetched a few lines above (and are fetched from the context created/refreshed in the calling function – `handleSyncResponse`),
* in an unmodified code (before https://github.com/duckduckgo/BrowserServicesKit/commit/b6fa81101e37b6de4cbbc304d962a00dca247c6e) fetching individual settings values is happening after fetching timestamps,
* hence we could get a lastModified that is still old (as far as a few-ms old context is concerned), and a new value,
* so I looked for a solution that brings fetching settings values as close to creating/refreshing the context as possible.

Ideally this could be done here in `handleSyncResponse` (but this may be hair-splitting):
```swift
context.performAndWait {
    while true {
        do {
            let originalValues: [Setting: String?] = try settingsHandlers.reduce(into: .init()) { partialResult, handler in
                partialResult[handler.key] = try handler.value.getValue()
            }
            let responseHandler = try SettingsResponseHandler( // ...
```

Reply:
> in an unmodified code (before https://github.com/duckduckgo/BrowserServicesKit/commit/b6fa81101e37b6de4cbbc304d962a00dca247c6e) fetching individual settings values is happening after fetching timestamps

I think array of settings were fetched, but not loaded into memory (by accessing property or unfaulting them in ay way), that's why I've assumed it is ok. But it is better to make it more explicit rather than rely on nuance when given object is being loaded into memory, especially as these operations should be pretty efficient. :)

Reply:
Thanks for the insight about faulting, it slipped my mind!